File: monsoon3.0/src/main/java/wtf/monsoon/impl/module/movement/NoSlow.java
package wtf.monsoon.impl.module.movement;

import io.github.nevalackin.homoBus.Listener;
import io.github.nevalackin.homoBus.annotations.EventLink;
import net.minecraft.block.BlockChest;
import net.minecraft.item.ItemSword;
import net.minecraft.network.play.client.*;
import net.minecraft.util.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.MovingObjectPosition;
import wtf.monsoon.Wrapper;
import wtf.monsoon.api.module.Category;
import wtf.monsoon.api.module.Module;
import wtf.monsoon.api.setting.Setting;
import wtf.monsoon.api.util.misc.PacketUtil;
import wtf.monsoon.api.util.misc.StringUtil;
import wtf.monsoon.impl.event.EventPacket;
import wtf.monsoon.impl.event.EventPostMotion;
import wtf.monsoon.impl.event.EventPreMotion;
import wtf.monsoon.impl.module.combat.Aura;

public class NoSlow extends Module {

    private final Setting<Mode> mode = new Setting<>("Mode", Mode.VANILLA)
            .describedBy("How to prevent slow down");

    private boolean sneakState = false;
    private boolean sprintState = false;

    public NoSlow() {
        super("No Slow", "Doesn't slow you down", Category.MOVEMENT);
        this.setMetadata(() -> StringUtil.formatEnum(mode.getValue()));
    }

    @Override
    public void onDisable() {
        super.onDisable();
        if (sneakState && !mc.thePlayer.isSneaking()) {
            PacketUtil.sendPacketNoEvent(new C0BPacketEntityAction(mc.thePlayer, C0BPacketEntityAction.Action.STOP_SNEAKING));
            sneakState = false;
        }

        if (sprintState && mc.thePlayer.isSprinting()) {
            PacketUtil.sendPacketNoEvent(new C0BPacketEntityAction(mc.thePlayer, C0BPacketEntityAction.Action.START_SPRINTING));
            sprintState = false;
        }
    }

    @EventLink
    public final Listener<EventPreMotion> eventPreMotionListener = e -> {
        switch (mode.getValue()) {
            case NCP:
                if(mc.thePlayer.isUsingItem() && player.isMoving()) {
                    if(mc.thePlayer.getHeldItem().getItem() instanceof ItemSword)
                        PacketUtil.sendPacketNoEvent(new C07PacketPlayerDigging(C07PacketPlayerDigging.Action.RELEASE_USE_ITEM, BlockPos.ORIGIN, EnumFacing.DOWN));
                    else
                        PacketUtil.sendPacketNoEvent(new C09PacketHeldItemChange(mc.thePlayer.inventory.currentItem));
                }
                break;
            case WATCHDOG:
                if (player.isMoving() && mc.thePlayer.isUsingItem())
                    PacketUtil.sendPacketNoEvent(new C09PacketHeldItemChange(mc.thePlayer.inventory.currentItem));
                break;
        }
    };

    @EventLink
    public final Listener<EventPostMotion> eventPostMotionListener = e -> {
        switch (mode.getValue()) {
            case NCP:
                if(mc.thePlayer.isUsingItem() && mc.thePlayer.getHeldItem().getItem() instanceof ItemSword)
                    PacketUtil.sendPacket(new C08PacketPlayerBlockPlacement(mc.thePlayer.getHeldItem()));
                break;
            case VANILLA:
                break;
        }
    };

    @EventLink
    public final Listener<EventPacket> eventPacketListener = e -> {
        if(mc.thePlayer == null || mc.theWorld == null) return;
        switch (mode.getValue()) {
            case WATCHDOG:
                /*MovingObjectPosition mop = mc.thePlayer.rayTrace(4, mc.thePlayer.ticksExisted);
                if(e.getPacket() instanceof C08PacketPlayerBlockPlacement) {
                    Aura aura = Wrapper.getModule(Aura.class);
                    if(mc.thePlayer.isUsingItem()
                            && mc.thePlayer.getHeldItem().getItem() instanceof ItemSword
                            && !(aura.isEnabled() && aura.getTarget() != null)
                            && !(mop.getBlockPos() != null && mc.theWorld.getBlockState(mop.getBlockPos()).getBlock() instanceof BlockChest)
                            && player.isMoving()) {
                        e.setCancelled(true);
                    }
                }*/
                break;
            case VANILLA:
                break;
        }
    };

    enum Mode {
        VANILLA("Vanilla"),
        WATCHDOG("Watchdog"),
        NCP("NCP");

        String modeName;

        Mode(String modeName) {
            this.modeName = modeName;
        }

        @Override
        public String toString() {
            return this.modeName;
        }
    }
}

File: monsoon-3.x-v2/src/main/java/wtf/monsoon/client/modules/player/NoSlow.kt
package wtf.monsoon.client.modules.player

import me.bush.eventbuskotlin.EventListener
import me.bush.eventbuskotlin.listener
import net.minecraft.item.ItemSword
import net.minecraft.network.play.client.C03PacketPlayer
import net.minecraft.network.play.client.C07PacketPlayerDigging
import net.minecraft.network.play.client.C08PacketPlayerBlockPlacement
import net.minecraft.network.play.client.C09PacketHeldItemChange
import net.minecraft.util.BlockPos
import net.minecraft.util.EnumFacing
import wtf.monsoon.Wrapper
import wtf.monsoon.backend.Category
import wtf.monsoon.backend.module.Module
import wtf.monsoon.backend.setting.Setting
import wtf.monsoon.client.event.EventMove
import wtf.monsoon.client.event.EventPacket
import wtf.monsoon.client.event.EventPostMotion
import wtf.monsoon.client.event.EventPreMotion
import wtf.monsoon.client.modules.movement.Flight
import wtf.monsoon.client.util.network.PacketUtil
import wtf.monsoon.client.util.math.getClosestMultipleOfDivisor

class NoSlow : Module("No Slow", "Take no fall damage", Category.PLAYER) {
    val mode = Setting<Mode>("Mode", Mode.VANILLA)

    @EventListener
    val preMotion = fun(_: EventPreMotion) {
        when(mode.getValue()) {
            Mode.NCP ->
                if(mc.thePlayer.isUsingItem && player.moving)
                    if(mc.thePlayer.heldItem.item is ItemSword)
                        PacketUtil.sendPacketNoEvent(C07PacketPlayerDigging(C07PacketPlayerDigging.Action.RELEASE_USE_ITEM, BlockPos.ORIGIN, EnumFacing.DOWN))

            Mode.ITEM_SWITCH_BUFFER ->
                if(mc.thePlayer.isUsingItem && player.moving)
                    PacketUtil.sendPacketNoEvent(C09PacketHeldItemChange(mc.thePlayer.inventory.currentItem))

            else -> {}
        }
    }

    @EventListener
    val postMotion = fun(e: EventPostMotion) {
        when(mode.getValue()) {
            Mode.NCP ->
                if(mc.thePlayer.isUsingItem && player.moving)
                    if(mc.thePlayer.heldItem.item is ItemSword)
                        PacketUtil.sendPacketNoEvent(C08PacketPlayerBlockPlacement(mc.thePlayer.heldItem))

            else -> {}
        }
    }

    fun shouldUseNoSlow(): Boolean {
        return this.isEnabled()
    }

    enum class Mode {
        VANILLA, NCP, ITEM_SWITCH_BUFFER
    }
}